functions/Export-SqlUser.ps1

```diff
@@ -250,50 +251,156 @@ https://dbatools.io/Export-SqlUser
                                 if ($DatabasePermission.PermissionState -eq "GrantWithGrant")
                                 {
                                     $WithGrant = "WITH GRANT OPTION"
+                                    $GrantDatabasePermission = 'GRANT'
                                 } 
                                 else 
                                 {
                                     $WithGrant = ""
+                                    $GrantDatabasePermission = $DatabasePermission.PermissionState.ToString().ToUpper()
                                 }
-                                $GrantDatabasePermission = $DatabasePermission.PermissionState.ToString().Replace("WithGrant", "").ToUpper()
 
-                                $outsql += "$($GrantDatabasePermission) $($DatabasePermission.PermissionType) TO [$($DatabasePermission.Grantee)] $WithGrant"
+                                $outsql += "$($GrantDatabasePermission) $($DatabasePermission.PermissionType) TO [$($DatabasePermission.Grantee)]$($WithGrant) AS [$($DatabasePermission.Grantor)];"
                             }
 
 
 	                        #Database Object Permissions
-                            foreach ($ObjectPermission in $db.EnumObjectPermissions() | Where-Object {@("sa","dbo","information_schema","sys") -notcontains $_.Grantee -and $_.Grantee -notlike "##*" -and $dbuser.Name -contains $_.Grantee})
+                            # NB: This is a bit of a mess for a couple of reasons
+                            # 1. $db.EnumObjectPermissions() doesn't enumerate all object types
+                            # 2. Some (x)Collection types can have EnumObjectPermissions() called
+                            #    on them directly (e.g. AssemblyCollection); others can't (e.g.
+                            #    ApplicationRoleCollection). Those that can't we iterate the
+                            #    collection explicitly and add each object's permission.
+
+                            $obj_perms = @()
```

Comment:
With regards to Performance, += rebuilds the object each time an item is added. It would be [a lot faster](https://foxdeploy.com/2016/03/23/coding-for-speed/) to use 

```
$perms = New-Object System.Collections.ArrayList
$null = $perms.Add($whatever)
```